proposed/0008-Connection_management.md

```diff
@@ -0,0 +1,212 @@
+- Feature Name: Improved Connection Management
+- Type: Enhancement
+- Related Components: routing, crust
+- Start Date: 23-09-2015
+- RFC PR: (leave this empty)
+- Issue Number: (leave this empty)
+
+# Summary
+
+Introduce the notion of clearly defined connection phases and establish directed routing network connections in conjunction with crust version 0.3 and later. The combined objective with crust proposed here for routing is to prepare for NAT traversal using UDP/uTP.
+
+# Motivation
+
+Starting with crust version 0.3, the API provides directional information on newly established connections.  This information is helpful for establishing a routing network with the correct topology. Already at bootstrapping it simplifies the logic for the routing node. No artificial distinction is needed for a "new bootstrap connection", as that depends on the state of the node and whether the node accepts or has connected to another node.
+
+# Detailed design
+
+## Rebootstrapping
+
+A first objective is the replacement of previous conditional variables with a clean flow through states.  This enables rebootstrapping, ie. when the connections to the network are lost, routing can automatically reconnect.  The states proposed in flow-order are `Disconnected`, `Bootstrapped`, `Relocated`, `Connected`, `GroupConnected`, `Terminated`.
+
+A full node will start at `Disconnected` and follow the states in the above order until `GroupConnected`. If a node would get disconnected then the cycle can restart at `Disconnected`.  Only when the user calls `::stop()`, does routing enter `Terminated`.  At `Terminated` the node remains active but does not accept new messages and terminates his connections to the network.
+
+A full node that cannot bootstrap (as such `Disconnected`), but accepts a connection from another node, will jump from `Disconnected` to `Relocated` by assigning itself a name.  Otherwise its behaviour is identical to normal behaviour.
```

Comment:
regarding `assigning itself a name`, according to Address Relocation rfc, isn't the name is `given` by the network group consensus?